### ISSUE TITLE
Is this expected ?

### ISSUE THREAD

**Issue by LucaGabi** [dart-lang/language#480]
```
if (v?.errors?.length > 0) { ... }
```

Error '>' called on null ?

**Comment by eernstg** [comment]
Yes, you're potentially comparing the null object to an argument (`0`), and `Null` does not have an `operator >`. So it's working as intended.

We're about to introduce a more flexible treatment of null-aware operators like `?.` (along with the introduction of non-nullable types), and this will allow you to use `v?.errors.length` because it won't try to get the `length` of null (it shortcuts as soon as null has been obtained at a `?.`). But I believe that it won't extend to binary expressions like `e1 > e2`. On the other hand, the compiler will tell you that `v?.errors.length` has a type which is potentially nullable and hence it's an error to compare it to anything with `>`, so the error will not occur, you just have to add a bit more code to handle it (like `(v?.errors.length ?? 0) > 0`).

The [feature specification on non-nullable types](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md) discusses null-aware operators in more detail [here](https://github.com/dart-lang/language/blob/master/accepted/future-releases/nnbd/feature-specification.md#null-aware-operator).

**Comment by LucaGabi** [comment]
Thanks .. I'll use this: (v?.errors.length ?? 0) > 0)

**Comment by eernstg** [comment]
This issue could be closed with a 'working as intended' outcome. But we have language issues like #394 ('do collection elements null short?') and #330 ('should null short-circuiting short-circuit spreads?') that are related to the discussion about how far to extend the null shortcuts.

So this issue could also be considered as similar, raising the issue about a possible extension of null shortcutting to cover more expressions than the ones currently in the feature specification, and in particular noting that we should investigate how well this has been working in C#.

@LucaGabi, it would be great if you could rephrase the [initial text of this issue](https://github.com/dart-lang/language/issues/480#issue-475048438) to propose such an extension, or close the issue if you don't want to push the idea.

**Comment by leafpetersen** [comment]
> Thanks .. I'll use this: (v?.errors.length ?? 0) > 0)

Note that in *current* Dart, this will still throw an exception if `v` is null, since `?.` is not "short-circuiting".  That is, `v?.errors.length` is currently treated as `((v == null) ? null : v.errors).length` (ignoring the duplicate evaluation of `v`).  In the upcoming NNBD release, we will be changing this so that it is treated as `(v == null) ? null : v.errors.length`.

So you may want to use `(v?.errors?.length ?? 0) > 0)` for now.

**Comment by leafpetersen** [comment]
> and in particular noting that we should investigate how well this has been working in C#.

I looked at this a bit.  I can't find the spec for it, but de facto there seems to be an implicit null converting overload defined for `>` , which seems to be why the above works.  The code below gets a warning that "Comparing with null of type int? always returns false", and does indeed result in "False" being printed.

```dart
public class Program
{
	public static String toString(object x) {
	if (x == null) return "null";
	return x.ToString();
}
	public static void Main()
	{
		Console.WriteLine(toString(null > 0));
	}
}
```

**Comment by eernstg** [comment]
> an implicit null converting overload defined for `>`

OK, that actually sounds like an error prone approach: If something is missing then the answer to `something > aValue` could be true or false, and I don't see any inherent bias toward false.

But we could still use this issue to handle discussions about possible adjustments to the propagation of null shortcuts.

The argument that we have previously given against including larger expressions (e.g., letting `e1 > e2` evaluate to null in case `e1` evaluates to null due to null shorting) is that it gets difficult to read the code when a large expression `e` (think: half a page) suddenly evaluates to null, skipping the execution of half of its subexpressions in ways that we can't easily discern by looking at `e`, because there's some null-aware operator deeply buried in the middle of `e`.

The current rules will null short a sequence of member accesses (`e?.m1.m2.m3` where each `mi` could be a getter invocation `.myGetter`, a method invocation `method(a, b)`, a setter invocation, etc). That's a reasonably simple rule to remember, and it proceeds in the same direction (we skip elements to the right of the `?`), and anything that breaks those rules-of-simplicity will make it harder to read.

**Comment by srawlins** [comment]
If we're going to keep this open to track a null shortcutting problem or idea, I suggest we change the title.

**Comment by eernstg** [comment]
I suggested that as well. ;-)

Done.
